### PR TITLE
roachtest: validate test name

### LINKED ISSUE #66182: roachtest: outputs corrupt TeamCity markers
```
[06:50:30]
[Step 2/3] ##teamcity[testStarted name='cancel/tpch/distsql/queries=[7 9 20 21]
[06:50:32]
[Step 2/3] Error while parsing TeamCity service message: Value should end with "'". Valid service message has a form of "##teamcity[messageName name1='escaped_value' name2='escaped_value']" where escaped_value uses substitutions: '->|', [->|[, ]->|], |->||, newline->|n
[06:50:30]
[Step 2/3] ,nodes=3' flowId='cancel/tpch/distsql/queries=[7 9 20 21],nodes=3']
```

https://github.com/cockroachdb/cockroach/blob/7ad39e6af943a9d4deca9ee767d786af0f23163b/pkg/cmd/roachtest/test_runner.go#L609-L611

and some others are not using

https://github.com/cockroachdb/cockroach/blob/334968badb9e8d7818d535fef4cc1fc4b1612791/pkg/cmd/roachtest/test.go#L477-L493

but they should.

### PR DESCRIPTION
**Pull request by cucaroach**
fixes #66182

### CODE REVIEW

**Review by stevendanna — COMMENTED** [review]
Nice!

**Review by tbg — APPROVED** [review]
Thanks Tommy!

When I first saw this I was worried about the following scenario:

- everything is good, all tests conform
- someone opens PR for a new roachtest `12350#)%*!*%#` (you get the idea)
- PR gets reviewed, and since it isn't run by CI, it passes and gets merged
- nightlies now fail (in the sense that no single roachtest will run)
- Tobias is sad

However, we do run a few roachtests in CI, and I think that is enough to catch this - just the mere existence of the roachtest will fail any attempt to run any roachtest, and we do run `acceptance/*`. So long story short - looks good to me!

I'm currently moving approximately every line in roachtest around over in https://github.com/cockroachdb/cockroach/pull/66430, just a heads up, not sure which of us will get in first. But should be easy to fix if there are merge conflicts.

File: pkg/cmd/roachtest/test_registry.go [hunk 1]
```diff
@@ -87,6 +87,10 @@ func (r *testRegistry) Add(spec testSpec) {
 
 // prepareSpec validates a spec and does minor massaging of its fields.
 func (r *testRegistry) prepareSpec(spec *testSpec) error {
+	if matched, err := regexp.MatchString("^[a-zA-Z0-9-_=/,]+$", spec.Name); err != nil || !matched {
+		return fmt.Errorf("%s: must specify valid Name", spec.Name)
```
Comment (tbg) [inline pkg/cmd/roachtest/test_registry.go:91]: You could print the regexp here (`regexp.QuoteMeta(pattern)`), that would help those who do come up with rejected test names. As is, they'll have to figure out what a valid name is.

**Review by cucaroach — COMMENTED** [review]
Well I did notice that "roachtest list" convienently calls prepareSpec on every test.   Is there a way to get CI to call roachtest list and fail if a new test trips the validation?   Ie before yahor's fix for cancel you'd see:

```
❯ bin/roachtest list
cancel/tpch/distsql/queries=[7 9 20 21],nodes=3: must specify valid Name
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66470#-:-McEfeQn3fg0jdWGLA4T:b-77fujs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/cmd/roachtest/test_registry.go, line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66470#-McENvnEr5Fiu4Wwhpql-r1-91:-McEjVjG5RbvpjAF2o5I:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/733880bda16b68bb813a0f8a241287f5c3f9c315/pkg/cmd/roachtest/test_registry.go#L91)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

You could print the regexp here (`regexp.QuoteMeta(pattern)`), that would help those who do come up with rejected test names. As is, they'll have to figure out what a valid name is.
</blockquote></details>

:+1:

**Review by tbg — COMMENTED** [review]
oh, and `roachtest run` doesn't do the same? I was hoping it would. Try something like this then:

```diff
diff --git a/build/teamcity-local-roachtest.sh b/build/teamcity-local-roachtest.sh
index 6cd0ea88a5..1f0b6519ff 100755
--- a/build/teamcity-local-roachtest.sh
+++ b/build/teamcity-local-roachtest.sh
@@ -21,6 +21,10 @@ tc_start_block "Compile roachprod/workload/roachtest"
 run build/builder.sh make bin/roachprod bin/workload bin/roachtest
 tc_end_block "Compile roachprod/workload/roachtest"
 
+tc_start_block "Checking that all roachtest names are valid"
+run build/builder.sh bin/roachtest list &> artifacts/roachtests-list.log || (cat artifacts/roachtests-list.log && false)
+tc_end_block "Checking that all roachtest names are valid"
+
 tc_start_block "Run local roachtests"
 # TODO(peter,dan): curate a suite of the tests that works locally.
 # NB: roachtest has its own teamcity output format and posts issues (whenever
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66470#-:-McEpGmFEcTvQYaW7vKf:b44ffrf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

**Review by cucaroach — COMMENTED** [review]
no it does, I just didn't know it, I think we're good.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66470#-:-McGcFb67wtieWaIiCZ_:b-gfxg3x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

**Pull request merged**